owlkettle/widgetdef.nim

```diff
@@ -82,6 +87,46 @@ proc redraw*(viewable: Viewable): bool =
     viewable.viewed = newWidget
     result = true
 
+proc hasRef*(stateRef: StateRef): bool = not stateRef.isNil() and not stateRef.state.isNil()
+
+proc newRef*(subscribers: varargs[proc(state: WidgetState) {.closure.}]): StateRef = 
+  var observers = initHashSet[proc(state: WidgetState)]()
+  for subscriber in subscribers:
+    observers.incl(subscriber)
+  return StateRef(observers: observers)
+
+proc unwrapInternalWidget*(stateRef: StateRef): GtkWidget =
+  if stateRef.isNil() or stateRef.state.isNil():
+    return nil.GtkWidget
+  
+  return stateRef.state.unwrapInternalWidget()
+
+proc setRef*(stateRef: StateRef, state: WidgetState) =
+  if stateRef.isNil():
+    return
```

Comment:
I don't see why you check arguments for nil everywhere. At least from a code style perspective this seems like a very uncommon choice in Nim.

Reply:
A nil check is always preferable over a segfault. The reason why I typically go for optionals when a specific instance actually is allowed to not exist (which is the case for stateRef instances, a widget may not have one) is to make handling of it explicit.

With a ref-type you never know. It blew up in my face enough times during development that just having the nil-checks there was the far safer option.

Reply:
The caller should check. Having the callee just silently do nothing is probably worse than a segfault, because if the stateRef is nil, that is a logic error.

Reply:
I'll stick to it as it's the code-style in the rest of the project, but generally I'd say I highly prefer Options due to their explicit nature of informing you "this might not exist", while in all other cases you just have to know that sth might not exist.